Mage.Sets/src/mage/cards/e/EdgarMasterMachinist.java

```diff
@@ -0,0 +1,208 @@
+        this.power = new MageInt(2);
+        this.toughness = new MageInt(4);
+
+        // Once during each of your turns, you may cast an artifact spell from your graveyard. If you cast a spell this way, that artifact enters tapped.
+        Ability castAbility = new SimpleStaticAbility(new EdgarMasterMachinistCastFromGraveyardEffect());
+        castAbility.setIdentifier(MageIdentifier.EdgarMasterMachinistWatcher);
+        castAbility.addWatcher(new EdgarMasterMachinistWatcher());
+        this.addAbility(castAbility);
+
+        // Tools -- Whenever Edgar attacks, it gets +X/+0 until end of turn, where X is the greatest mana value among artifacts you control.
+        this.addAbility(new AttacksTriggeredAbility(
+                new BoostSourceEffect(GreatestAmongPermanentsValue.MANAVALUE_CONTROLLED_ARTIFACTS, StaticValue.get(0), Duration.EndOfTurn, "it")
+        ).withFlavorWord("Tools").addHint(GreatestAmongPermanentsValue.MANAVALUE_CONTROLLED_ARTIFACTS.getHint()));
+    }
+
+    private EdgarMasterMachinist(final EdgarMasterMachinist card) {
+        super(card);
+    }
+
+    @Override
+    public EdgarMasterMachinist copy() {
+        return new EdgarMasterMachinist(this);
+    }
+}
+
+// Does double duty:
+// - Keep watch of the "Once during each of your turns" usage of the ability, similar to CastFromGraveyardOnceEachTurnAbility
+// - Watcher altering spells being cast with the right MageIdentifier.
+//   Similar to Intrepid Paleontologist and the like for entering with counters.
+class EdgarMasterMachinistWatcher extends Watcher {
+
+    // set of each Edgar having used their ability this turn.
+    private final Set<MageObjectReference> usedFrom = new HashSet<>();
+
+    EdgarMasterMachinistWatcher() {
+        super(WatcherScope.GAME);
+    }
+
+    @Override
+    public void watch(GameEvent event, Game game) {
+
+        if (GameEvent.EventType.SPELL_CAST.equals(event.getType())
+                && event.hasApprovingIdentifier(MageIdentifier.EdgarMasterMachinistWatcher)) {
+            // register a spell was cast using Edgar ability
+            usedFrom.add(event.getApprovingObject().getApprovingMageObjectReference());
+
+            // Adds the continuous effect for the cast artifact to enter tapped
+            Spell target = game.getSpell(event.getTargetId());
+            if (target != null) {
+                MageObjectReference mor = new MageObjectReference(target, game);
+                // Not that happy with setting the source of the ContinuousEffect to the SpellAbility
+                game.getState().addEffect(new EdgarMasterMachinistEntersTappedEffect(mor), target.getSpellAbility());
+            }
+        }
+    }
+
+    @Override
+    public void reset() {
+        super.reset();
+        usedFrom.clear();
+    }
+
+    boolean abilityUsed(MageObjectReference mor) {
+        return usedFrom.contains(mor);
+    }
+}
+
+class EdgarMasterMachinistCastFromGraveyardEffect extends AsThoughEffectImpl {
```

Comment:
It's better to improve `CastFromGraveyardOnceEachTurnAbility` to setup custom mage identifier instead copy whole ability.